src/lib/transition_frontier/full_frontier/full_frontier.ml

```diff
@@ -0,0 +1,566 @@
+                 [ ( "selection_context"
+                   , `String "comparing new breadcrumb to best tip" ) ])
+          = `Take
+        then Full.E.E (Best_tip_changed breadcrumb_hash) :: diffs
+        else diffs
+      in
+      (* reverse diffs so that they are applied in the correct order *)
+      List.rev diffs )
+
+(* TODO: refactor metrics tracking outside of apply_diff (could maybe even be an extension?) *)
+let apply_diff (type mutant) t (diff : (Diff.full, mutant) Diff.t) :
+    mutant * State_hash.t option =
+  match diff with
+  | New_node (Full breadcrumb) ->
+      let breadcrumb_hash = Breadcrumb.state_hash breadcrumb in
+      let parent_hash = Breadcrumb.parent_hash breadcrumb in
+      let parent_node = Hashtbl.find_exn t.table parent_hash in
+      Hashtbl.add_exn t.table ~key:breadcrumb_hash
+        ~data:{breadcrumb; successor_hashes= []; length= parent_node.length + 1} ;
+      Hashtbl.set t.table ~key:parent_hash
+        ~data:
+          { parent_node with
+            successor_hashes= breadcrumb_hash :: parent_node.successor_hashes
+          } ;
+      Coda_metrics.(Gauge.inc_one Transition_frontier.active_breadcrumbs) ;
+      Coda_metrics.(Counter.inc_one Transition_frontier.total_breadcrumbs) ;
+      ((), None)
+  | Best_tip_changed new_best_tip ->
+      let old_best_tip = t.best_tip in
+      t.best_tip <- new_best_tip ;
+      (old_best_tip, None)
+  | Root_transitioned
+      {new_root= {hash= new_root_hash; _}; garbage= Full garbage_breadcrumbs}
+    ->
+      (* The transition frontier at this point in time has the following mask topology:
+       *
+       *   (`s` represents a snarked ledger, `m` represents a mask)
+       * 
+       *     garbage
+       *     [m...]
+       *       ^
+       *       |          successors
+       *       m0 -> m1 -> [m...]
+       *       ^
+       *       |
+       *       s
+       *
+       * In this diagram, the old root's mask (`m0`) is parented off of the root snarked
+       * ledger database, and the new root's mask (`m1`) is parented off of the `m0`.
+       * There is also some garbage parented off of `m0`, and some successors that will
+       * be kept in the tree after transition which are parented off of `m1`.
+       *
+       * In order to move the root, we must form a mask `m1'` with the same merkle root
+       * as `m1`, except that it is parented directly off of the root snarked ledger
+       * instead of `m0`. Furthermore, the root snarked ledger `s` may update to another
+       * merkle root as `s'` if there is a proof emitted in the transition between `m0`
+       * and `m1`.
+       *
+       * To form a mask `m1'` and update the snarked ledger from `s` to `s'` (which is a
+       * noop in the case of no ledger proof emitted between `m0` and `m1`), we must perform
+       * the following operations on masks in order:
+       *
+       *     1) unattach and destroy all the garbage (to avoid unecessary trickling of
```

Comment:
can you add the steps inline in the function like the old implementation -- it helps keep track of what is going on here.